In [1]:
import jupyter_black

jupyter_black.load()

These tutorials are written to help you get familiar with some of the common functionalities that most actuaries can use in their day-to-day responsibilities that are provided by the `chainladder` package. We will also be using the datasets that already come included with the package, allowing you to follow and reproduce the results as shown here.

Keep in mind that these tutorials were written to only demonstrate the functionalities of the package, and the user should always follow all applicable laws, the Code of Professional Conduct, applicable Actuarial Standards of Practice, and exercise their best actuarial judgement. These tutorials are not written in a way that encourage certain workflow, or recommendation, when it comes to analyzing a dataset or rendering an actuarial opinion.

The tutorials assume that you have the basic understanding of commonly used actuarial terms, and can independently perform an actuarial analysis in another tool, such as Microsoft Excel or another actuarial software. Furthermore, it is assumed that you already have some familiarity with Python, and that you have the basic knowledge and experience in using some common packages that are popular in the Python community, such as `pandas` and `numpy`.

All tutorials and exercises rely on `chainladder` v0.8.18 and later. If you have trouble reconciling the results from your workflow to this tutorial, you should verify the versions of the packages installed in your work environment and check the release notes in case updates patches are issued subsequently.

In [2]:
import pandas as pd
import numpy as np
import chainladder as cl

print("pandas: " + pd.__version__)
print("numpy: " + np.__version__)
print("chainladder: " + cl.__version__)

Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
Intel MKL WARNING: Support of Intel(R) Streaming SIMD Extensions 4.2 (Intel(R) SSE4.2) enabled only processors has been deprecated. Intel oneAPI Math Kernel Library 2025.0 will require Intel(R) Advanced Vector Extensions (Intel(R) AVX) instructions.
pandas: 2.1.4
numpy: 1.26.3
chainladder: 0.8.18


# Working with Triangles

## Importing Data

Let's begin by looking at a raw triangle dataset and load it into a `pandas.DataFrame`. We'll use the data `raa`, which is available from the repository. Note that this dataset is currently in `csv` format.

In [3]:
raa_df = pd.read_csv(
    "https://raw.githubusercontent.com/casact/chainladder-python/master/chainladder/utils/data/raa.csv"
)
raa_df.head(20)

development  origin   values
0          1981    1981   5012.0
1          1982    1982    106.0
2          1983    1983   3410.0
3          1984    1984   5655.0
4          1985    1985   1092.0
5          1986    1986   1513.0
6          1987    1987    557.0
7          1988    1988   1351.0
8          1989    1989   3133.0
9          1990    1990   2063.0
10         1982    1981   8269.0
11         1983    1982   4285.0
12         1984    1983   8992.0
13         1985    1984  11555.0
14         1986    1985   9565.0
15         1987    1986   6445.0
16         1988    1987   4020.0
17         1989    1988   6947.0
18         1990    1989   5395.0
19         1983    1981  10907.0

The dataset has three columns: 
* development: or valuation time, in this case, the valuation year
* origin: or accident date, in this case, the accident year
* values: the values recorded for the specific accident date at the specific valuation time (such as incurred losses, paid losses, or claim counts), in this case, these are just "values" within the triangle, and has no specific metrics unit associated with them

A table of loss experience showing total losses for a certain period (origin) at various, regular valuation dates (development), reflects the change in amounts as claims mature and emerge. Older periods in the table will have one more entry than the next youngest period, leading to the triangle shape of the data in the table or any other measure that matures over time from an origin date. Loss triangles can be used to determine loss development for a given risk.

Let's put our data into the `chainladder.Triangle` format.

In [4]:
raa = cl.Triangle(
    data=raa_df,
    origin="origin",
    development="development",
    columns="values",
    cumulative=True,
)
raa

12       24       36       48       60       72       84       96       108      120
1981  5012.0   8269.0  10907.0  11805.0  13539.0  16181.0  18009.0  18608.0  18662.0  18834.0
1982   106.0   4285.0   5396.0  10666.0  13782.0  15599.0  15496.0  16169.0  16704.0      NaN
1983  3410.0   8992.0  13873.0  16141.0  18735.0  22214.0  22863.0  23466.0      NaN      NaN
1984  5655.0  11555.0  15766.0  21266.0  23425.0  26083.0  27067.0      NaN      NaN      NaN
1985  1092.0   9565.0  15836.0  22169.0  25955.0  26180.0      NaN      NaN      NaN      NaN
1986  1513.0   6445.0  11702.0  12935.0  15852.0      NaN      NaN      NaN      NaN      NaN
1987   557.0   4020.0  10946.0  12314.0      NaN      NaN      NaN      NaN      NaN      NaN
1988  1351.0   6947.0  13112.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1989  3133.0   5395.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1990  2063.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN

In the above example,
* `data` is the single `DataFrame` that contains columns representing all other arguments to the Triangle constructor. In our example, the dataset `raa_df`.
* `origin` is the representation of the accident, reporting or more generally the origin period of the triangle that will map to the `origin` dimension. In our example, the `origin` column
* `development` is the representation of the development/valuation periods of the triangle that will map to the `development` dimension. In our example, the `development` column.
* `columns` is the representation of the numeric data of the triangle that will map to the `columns` dimension. If `None`, then a single 'Total' key will be generated. In our example, the `values` column.
* `columuative` is the indicator of whether the triangle is cumulative or incremental. In our example, while it is not super obvious from looking at the raw data, our triangle dataset is actually a cumulative triangle. So we'll set this to `True`.
  
## Triangles Attributes

Now that we have our `Triangle` object declared within the `chainladder` package, we can get a lot of its attributes. First, let's get the latest diagonal of the Triangle with `.latest_diagonal`.

In [5]:
raa.latest_diagonal

1990
1981  18834.0
1982  16704.0
1983  23466.0
1984  27067.0
1985  26180.0
1986  15852.0
1987  12314.0
1988  13112.0
1989   5395.0
1990   2063.0

Another attribute that is commonly used is `.link_ratio` to get the LDFs of the triangle.

In [6]:
raa.link_ratio

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
1981   1.649840  1.319023  1.082332  1.146887  1.195140  1.112972  1.033261  1.002902  1.009217
1982  40.424528  1.259277  1.976649  1.292143  1.131839  0.993397  1.043431  1.033088       NaN
1983   2.636950  1.542816  1.163483  1.160709  1.185695  1.029216  1.026374       NaN       NaN
1984   2.043324  1.364431  1.348852  1.101524  1.113469  1.037726       NaN       NaN       NaN
1985   8.759158  1.655619  1.399912  1.170779  1.008669       NaN       NaN       NaN       NaN
1986   4.259749  1.815671  1.105367  1.225512       NaN       NaN       NaN       NaN       NaN
1987   7.217235  2.722886  1.124977       NaN       NaN       NaN       NaN       NaN       NaN
1988   5.142117  1.887433       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1989   1.721992       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

Another useful feature is the `.heatmap()` method.

In [7]:
raa.link_ratio.heatmap()

Here are some other attributes that might be useful to the user:
* `is_cumulative`: returns True if the data across the development periods is cumulative, or False if it is incremental.
* `is_ultimate`: returns True if the ultimate values are contained in the triangle.
* `is_val_tri`: returns True if the development period is stated as a valuation data as opposed to an age, i.e. Schedule P style triangle (True) or the more commonly used triangle by development age (False).
* `is_full`: returns True if the triangle has been "squared".

In [8]:
print("Is triangle cumulative?", raa.is_cumulative)
print("Does triangle contain ultimate projections?", raa.is_ultimate)
print("Is this a valuation triangle?", raa.is_val_tri)
print('Has the triangle been "squared"?', raa.is_full)

Is triangle cumulative? True
Does triangle contain ultimate projections? False
Is this a valuation triangle? False
Has the triangle been "squared"? False


We can also inspect the triangle to understand its data granularity with `origin_grain` and `development_grain`. The supported `grains` are:
* monthly: denoted with `M`
* quarterly: denoted with `Q`
* semi-annually: denoted with `S`
* annually: denoted with `Y`

In [9]:
print("Origin grain:", raa.origin_grain)
print("Development grain:", raa.development_grain)

Origin grain: Y
Development grain: Y


## Manipulating Triangles
There are also useful methods to convert an cumulative triangle into an incremental one with `.cum_to_incr()`.

In [10]:
raa.cum_to_incr()

12      24      36      48      60      72      84     96     108    120
1981  5012.0  3257.0  2638.0   898.0  1734.0  2642.0  1828.0  599.0   54.0  172.0
1982   106.0  4179.0  1111.0  5270.0  3116.0  1817.0  -103.0  673.0  535.0    NaN
1983  3410.0  5582.0  4881.0  2268.0  2594.0  3479.0   649.0  603.0    NaN    NaN
1984  5655.0  5900.0  4211.0  5500.0  2159.0  2658.0   984.0    NaN    NaN    NaN
1985  1092.0  8473.0  6271.0  6333.0  3786.0   225.0     NaN    NaN    NaN    NaN
1986  1513.0  4932.0  5257.0  1233.0  2917.0     NaN     NaN    NaN    NaN    NaN
1987   557.0  3463.0  6926.0  1368.0     NaN     NaN     NaN    NaN    NaN    NaN
1988  1351.0  5596.0  6165.0     NaN     NaN     NaN     NaN    NaN    NaN    NaN
1989  3133.0  2262.0     NaN     NaN     NaN     NaN     NaN    NaN    NaN    NaN
1990  2063.0     NaN     NaN     NaN     NaN     NaN     NaN    NaN    NaN    NaN

You can also convert an incremental triangle to a cumuative one with `.incr_to_cum()`.

In [11]:
raa.cum_to_incr().incr_to_cum()

12       24       36       48       60       72       84       96       108      120
1981  5012.0   8269.0  10907.0  11805.0  13539.0  16181.0  18009.0  18608.0  18662.0  18834.0
1982   106.0   4285.0   5396.0  10666.0  13782.0  15599.0  15496.0  16169.0  16704.0      NaN
1983  3410.0   8992.0  13873.0  16141.0  18735.0  22214.0  22863.0  23466.0      NaN      NaN
1984  5655.0  11555.0  15766.0  21266.0  23425.0  26083.0  27067.0      NaN      NaN      NaN
1985  1092.0   9565.0  15836.0  22169.0  25955.0  26180.0      NaN      NaN      NaN      NaN
1986  1513.0   6445.0  11702.0  12935.0  15852.0      NaN      NaN      NaN      NaN      NaN
1987   557.0   4020.0  10946.0  12314.0      NaN      NaN      NaN      NaN      NaN      NaN
1988  1351.0   6947.0  13112.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1989  3133.0   5395.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1990  2063.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN

Another useful one is to convert a development triangle to a valuation triangle (Schedule P style), with `.dev_to_val()`.

In [12]:
raa.dev_to_val()

1981    1982     1983     1984     1985     1986     1987     1988     1989     1990
1981  5012.0  8269.0  10907.0  11805.0  13539.0  16181.0  18009.0  18608.0  18662.0  18834.0
1982     NaN   106.0   4285.0   5396.0  10666.0  13782.0  15599.0  15496.0  16169.0  16704.0
1983     NaN     NaN   3410.0   8992.0  13873.0  16141.0  18735.0  22214.0  22863.0  23466.0
1984     NaN     NaN      NaN   5655.0  11555.0  15766.0  21266.0  23425.0  26083.0  27067.0
1985     NaN     NaN      NaN      NaN   1092.0   9565.0  15836.0  22169.0  25955.0  26180.0
1986     NaN     NaN      NaN      NaN      NaN   1513.0   6445.0  11702.0  12935.0  15852.0
1987     NaN     NaN      NaN      NaN      NaN      NaN    557.0   4020.0  10946.0  12314.0
1988     NaN     NaN      NaN      NaN      NaN      NaN      NaN   1351.0   6947.0  13112.0
1989     NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN   3133.0   5395.0
1990     NaN     NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN   2063.0

And of course, you can convert it back with `.val_to_dev()`.

In [13]:
raa.dev_to_val().val_to_dev()

12       24       36       48       60       72       84       96       108      120
1981  5012.0   8269.0  10907.0  11805.0  13539.0  16181.0  18009.0  18608.0  18662.0  18834.0
1982   106.0   4285.0   5396.0  10666.0  13782.0  15599.0  15496.0  16169.0  16704.0      NaN
1983  3410.0   8992.0  13873.0  16141.0  18735.0  22214.0  22863.0  23466.0      NaN      NaN
1984  5655.0  11555.0  15766.0  21266.0  23425.0  26083.0  27067.0      NaN      NaN      NaN
1985  1092.0   9565.0  15836.0  22169.0  25955.0  26180.0      NaN      NaN      NaN      NaN
1986  1513.0   6445.0  11702.0  12935.0  15852.0      NaN      NaN      NaN      NaN      NaN
1987   557.0   4020.0  10946.0  12314.0      NaN      NaN      NaN      NaN      NaN      NaN
1988  1351.0   6947.0  13112.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1989  3133.0   5395.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1990  2063.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN

## 4-D Triangle

The triangle described so far is a two-dimensional (accident date by valuation date) structure that spans multiple cells of data. This is a useful structure for exploring individual triangles, but becomes more problematic when working with **sets** of triangles. `Pandas` does not have a triangle `dtype`, but if it did, working with sets of triangles would be much more convenient. To facilitate working with more than one triangle at a time,  `chainladder.Triangle` acts like a `pandas.DataFrame` (with an index and columns) where each cell (row x col) is an individual triangle. This structure manifests itself as a four-dimensional space. Let's take a look at another sample dataset, `clrd`.

In [14]:
clrd_df = pd.read_csv(
    "https://raw.githubusercontent.com/casact/chainladder-python/master/chainladder/utils/data/clrd.csv"
)
clrd_df.head()

GRCODE               GRNAME  AccidentYear  DevelopmentYear  DevelopmentLag  \
0      86  Allstate Ins Co Grp          1988             1988               1   
1      86  Allstate Ins Co Grp          1988             1989               2   
2      86  Allstate Ins Co Grp          1988             1990               3   
3      86  Allstate Ins Co Grp          1988             1991               4   
4      86  Allstate Ins Co Grp          1988             1992               5   

   IncurLoss  CumPaidLoss  BulkLoss  EarnedPremDIR  EarnedPremCeded  \
0     367404        70571    127737         400699             5957   
1     362988       155905     60173         400699             5957   
2     347288       220744     27763         400699             5957   
3     330648       251595     15280         400699             5957   
4     354690       274156     27689         400699             5957   

   EarnedPremNet  Single  PostedReserve97     LOB  
0         394742       0           281872  wkcomp  
1         394742       0           281872  wkcomp  
2         394742       0           281872  wkcomp  
3         394742       0           281872  wkcomp  
4         394742       0           281872  wkcomp

Let's load the data into the sets of triangles.

In [15]:
clrd = cl.Triangle(
    data=clrd_df,
    origin="AccidentYear",
    development="DevelopmentYear",
    columns=[
        "IncurLoss",
        "CumPaidLoss",
        "BulkLoss",
        "EarnedPremDIR",
        "EarnedPremCeded",
        "EarnedPremNet",
    ],
    index=["GRNAME", "LOB"],
    cumulative=True,
)
clrd

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                       (775, 6, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

In this example, `data`, `origin`, `development`, and `cumulative` are all no different from what we had done before. But, we need to use `columns` a bit differently, and use declare a new variable, `index`.

* `columns` is the list of "triangles" metrics that we will have. Think of this as the type of triangle metrics that we can use to describe a segment. e.g. Paid, Incurred, Closed Claim counts, or even exposure data. Note that even exposure data do not develop over time, they can still be presented in a triangle format. In our example, `IncurLoss` for financial incurred loss (not just reported, as in paid + case + IBNR), `CumPaidLoss` for paid loss, `BulkLoss` for case reserves, `EarnedPremDIR` for direct and assumed premium earned, `EarnedPremCeded` for ceded premium earned, and `EarnedPremNet` for the net premium earned.
* `index` is our portfolio segments. In our example, a combination of `GRNAME`, the Company Name, and `LOB`, the line of business.

Since 4D structures do not fit nicely on 2D screens, we see a summary view instead that describes the structure rather than the underlying data itself. 

We see 5 rows of information:
* Valuation: the valuation date.
* Grain: the granularity of the data, `O` stands for origin, and `D` stands for development, `OYDY` represents triangles with accident year by development year.
* Shape: contains 4 numbers, represents the 4-D structure. This sample triangle represents a collection of 775x6 or 4,650 triangles that are themselves 10 accident years by 10 development periods.
    * 775: the number of segments, which is the combination of `index`, that represents the data segments. In this case, it is each of the `GRNAME` and `LOB` combination.
    * 6: the number of triangles for each segment, which is also the columns `[IncurLoss, CumPaidLoss, BulkLoss, EarnedPremDIR, EarnedPremCeded, EarnedPremNet]`.
    * 10: the number of accident periods.
    * 10: the number of valuation periods.
* Index: the segmentation level of the triangles.
* Columns: the value types recorded in the triangles.

Now we have a 4D triangle, let's do some `pandas`-style operations. First, we can filter.

In [16]:
clrd[clrd["LOB"] == "wkcomp"]

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                       (132, 6, 10, 10)
Index:                                          [GRNAME, LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

Note that only the shape changed, from `(775, 6, 10, 10)` to `(132, 6, 10, 10)`.

Next, you can use `.loc` to filter by index name.

In [17]:
clrd.loc["Allstate Ins Co Grp"]

Triangle Summary
Valuation:                                            1997-12
Grain:                                                   OYDY
Shape:                                         (2, 6, 10, 10)
Index:                                                  [LOB]
Columns:    [IncurLoss, CumPaidLoss, BulkLoss, EarnedPremD...

Let's see what LOB `Allstate Ins Co Grp` writes.

In [18]:
clrd.loc["Allstate Ins Co Grp"].index

LOB
0  prodliab
1    wkcomp

Since we have `.loc`, we must also have `.iloc` by index location. You can even chain them together. 

Let's get `Allstate Ins Co Grp`'s `prodliab` by calling `iloc[0]` and get the `CumPaidLoss` triangle.

In [19]:
clrd.loc["Allstate Ins Co Grp"].iloc[0]["CumPaidLoss"]

12       24       36       48       60       72       84       96       108      120
1988  1501.0   3916.0   8834.0  17450.0  22495.0  28687.0  31311.0  32039.0  36357.0  36358.0
1989  1697.0   5717.0  10442.0  18125.0  23284.0  30092.0  34338.0  41094.0  41164.0      NaN
1990  1373.0   4002.0  10829.0  16695.0  21788.0  25332.0  34875.0  34893.0      NaN      NaN
1991  1069.0   4594.0   6920.0   9996.0  13249.0  19221.0  19256.0      NaN      NaN      NaN
1992  1134.0   3068.0   5412.0   8210.0  19164.0  19187.0      NaN      NaN      NaN      NaN
1993   979.0   3079.0   6407.0  16113.0  16131.0      NaN      NaN      NaN      NaN      NaN
1994  1397.0   2990.0  25688.0  26030.0      NaN      NaN      NaN      NaN      NaN      NaN
1995  1016.0  21935.0  22095.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1996  9852.0  10071.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
1997   319.0      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN

`iloc[...]` actually can take in 4 parameters, index, columns, origin, and development. For example, if we want `Allstate Ins Co Grp`'s `prodliab` index, we can search for it first, then call the indices of `CumPaidLoss`, and origin `1990`.

In [20]:
clrd.index[clrd.index["GRNAME"] == "Allstate Ins Co Grp"]

GRNAME       LOB
21  Allstate Ins Co Grp  prodliab
22  Allstate Ins Co Grp    wkcomp

In [21]:
clrd.iloc[21, 1, 2, :]

12      24       36       48       60       72       84       96   108  120
1990  1373.0  4002.0  10829.0  16695.0  21788.0  25332.0  34875.0  34893.0  NaN  NaN

We can also use other `pandas` filter functions, for example, getting the four `CumPaidLoss` diagonals between 1990 and 1993 wit hthe help of `.valuation`.

In [22]:
paid_tri = clrd.loc["Allstate Ins Co Grp"].iloc[0]["CumPaidLoss"]
paid_tri[(paid_tri.valuation >= "1990") & (paid_tri.valuation < "1994")]["CumPaidLoss"]

12      24       36       48       60       72
1988     NaN     NaN   8834.0  17450.0  22495.0  28687.0
1989     NaN  5717.0  10442.0  18125.0  23284.0      NaN
1990  1373.0  4002.0  10829.0  16695.0      NaN      NaN
1991  1069.0  4594.0   6920.0      NaN      NaN      NaN
1992  1134.0  3068.0      NaN      NaN      NaN      NaN
1993   979.0     NaN      NaN      NaN      NaN      NaN

Another commonly used filter is `.development`, let's get the three columns between age 36 and age 60.

In [23]:
paid_tri[(paid_tri.development >= 36) & (paid_tri.development <= 60)]

36       48       60
1988   8834.0  17450.0  22495.0
1989  10442.0  18125.0  23284.0
1990  10829.0  16695.0  21788.0
1991   6920.0   9996.0  13249.0
1992   5412.0   8210.0  19164.0
1993   6407.0  16113.0  16131.0
1994  25688.0  26030.0      NaN
1995  22095.0      NaN      NaN
1996      NaN      NaN      NaN
1997      NaN      NaN      NaN

With complete flexibility in the ability to slice subsets of triangles, we can use basic arithmetic to derive new triangles, which is commonly used as diagnostics to explore trends.

In [24]:
clrd["CaseIncurLoss"] = clrd["IncurLoss"] - clrd["BulkLoss"]
clrd["CaseIncurLoss"]

Triangle Summary
Valuation:           1997-12
Grain:                  OYDY
Shape:      (775, 1, 10, 10)
Index:         [GRNAME, LOB]
Columns:     [CaseIncurLoss]

Note that even though `clrd["CaseIncurLoss"]` is declared as a new variable, it actually comes with all 775 "indexes", i.e. we have 775 `clrd["CaseIncurLoss"]` triangles. But we can use `.sum()` to see the sum of them.

In [25]:
clrd["CaseIncurLoss"].sum()

12          24          36          48          60          72          84          96          108         120
1988   7778398.0   9872876.0  10537707.0  10973808.0  11175391.0  11265524.0  11288288.0  11305023.0  11323995.0  11327627.0
1989   8734319.0  10844720.0  11822136.0  12279311.0  12481505.0  12567543.0  12608487.0  12633539.0  12639258.0         NaN
1990   9325252.0  11913461.0  12985113.0  13459843.0  13646077.0  13718445.0  13755879.0  13768960.0         NaN         NaN
1991   9564486.0  12159826.0  13216383.0  13659541.0  13821032.0  13903084.0  13964163.0         NaN         NaN         NaN
1992  10539619.0  13125930.0  14120971.0  14563964.0  14755405.0  14850140.0         NaN         NaN         NaN         NaN
1993  11402448.0  14043343.0  15095232.0  15576086.0  15775057.0         NaN         NaN         NaN         NaN         NaN
1994  12411107.0  15005424.0  16095699.0  16650937.0         NaN         NaN         NaN         NaN         NaN         NaN
1995  12686394.0  15140099.0  16223016.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN
1996  12627293.0  14956778.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN
1997  12705993.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN

Let's look at the (sum of) Paid to (sum of) Incurred ratio triangle. Does it look like the ratios are changing over time? Using `.heatmap()` usually helps with spotting trends.

In [26]:
(clrd["CumPaidLoss"].sum() / clrd["CaseIncurLoss"].sum()).heatmap()

Compare the result to `(clrd["CumPaidLoss"] / clrd["CaseIncurLoss"]).sum()`, which looks odd, can you figure out why? This is because we are summing all of the quotients of paid losses over incurred losses at each index.

In [27]:
(clrd["CumPaidLoss"] / clrd["CaseIncurLoss"]).sum()

12          24          36          48          60          72          84          96          108        120
1988  183.174954  292.568472  331.870839  398.920441  494.883976  473.435946  485.361460  490.862840  489.800188  491.31401
1989  200.685859  303.319078  373.150572  420.302929  463.582968  484.426264  497.129633  505.378356  503.292688        NaN
1990  215.342110  369.604469  378.755996  476.510210  518.929698  505.843304  513.896098  520.722478         NaN        NaN
1991  164.685591  327.687927  396.143934  444.785517  485.415433  503.483137  514.193116         NaN         NaN        NaN
1992  218.254666  340.385259  411.630112  476.672321  499.918913  516.520918         NaN         NaN         NaN        NaN
1993  231.650900  352.995731  423.386563  483.572803  517.404169         NaN         NaN         NaN         NaN        NaN
1994  235.049595  355.944571  436.773937  498.180201         NaN         NaN         NaN         NaN         NaN        NaN
1995  235.370730  369.357522  445.787686         NaN         NaN         NaN         NaN         NaN         NaN        NaN
1996  245.990104  384.577758         NaN         NaN         NaN         NaN         NaN         NaN         NaN        NaN
1997  268.778113         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN        NaN

## Triangle Adjustments

Another adjustment we can make to the triangle is to apply a trend. We can do that by calling `chainladder.Trend()`, which is actually an estimator. It takes in a few variables:
- `trends`: the list containing the annual trends expressed as a decimal. For example, 5% decrease should be stated as -0.05.
- `dates`: a list-like of (start, end) dates to correspond to the trend list.
- `axis` (options: [‘origin’, ‘valuation’]): the axis on which to apply the trend.

Let's say we want a 5% trend from `1992-12-31` to `1991-01-01`. You can then call `.trend_` attribute to view the trend factors.

In [28]:
cl.Trend(trends=[0.05], dates=[("1992-12-31", "1991-01-01")], axis="origin").fit(
    clrd["CumPaidLoss"].sum()
).trend_

12        24        36        48        60        72        84        96        108       120
1988  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027
1989  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027       NaN
1990  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027       NaN       NaN
1991  1.050000  1.050000  1.050000  1.050000  1.050000  1.050000  1.050000       NaN       NaN       NaN
1992  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN
1993  1.000000  1.000000  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN
1994  1.000000  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN       NaN
1995  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1996  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1997  1.000000       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

Multipart trend is also possible, since `trends` and `dates` can accept lists.

In [29]:
cl.Trend(
    trends=[0.05, -0.10],
    dates=[("1992-12-31", "1991-01-01"), ("1990-12-31", "1989-01-01")],
    axis="origin",
).fit(clrd["CumPaidLoss"].sum()).trend_

12        24        36        48        60        72        84        96        108       120
1988  0.897245  0.897245  0.897245  0.897245  0.897245  0.897245  0.897245  0.897245  0.897245  0.897245
1989  0.988224  0.988224  0.988224  0.988224  0.988224  0.988224  0.988224  0.988224  0.988224       NaN
1990  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027  1.098027       NaN       NaN
1991  1.050000  1.050000  1.050000  1.050000  1.050000  1.050000  1.050000       NaN       NaN       NaN
1992  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN
1993  1.000000  1.000000  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN
1994  1.000000  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN       NaN
1995  1.000000  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1996  1.000000  1.000000       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN
1997  1.000000       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN       NaN

`chainladder.Triangle()` objects are awesome, but what if you need to get back out to `pandas`? `.to_frame()` is a very handy method to know. It converts the `chainladder.Triangle()` objects back to a `pandas.DataFrame()` object.

In [30]:
clrd["CumPaidLoss"].sum().to_frame()

12          24          36          48          60   \
1988-01-01  3577780.0   7059966.0   8826151.0   9862687.0  10474698.0   
1989-01-01  4090680.0   7964702.0   9937520.0  11098588.0  11766488.0   
1990-01-01  4578442.0   8808486.0  10985347.0  12229001.0  12878545.0   
1991-01-01  4648756.0   8961755.0  11154244.0  12409592.0  13092037.0   
1992-01-01  5139142.0   9757699.0  12027983.0  13289485.0  13992821.0   
1993-01-01  5653379.0  10599423.0  12953812.0  14292516.0  15005138.0   
1994-01-01  6246447.0  11394960.0  13845764.0  15249326.0         NaN   
1995-01-01  6473843.0  11612151.0  14010098.0         NaN         NaN   
1996-01-01  6591599.0  11473912.0         NaN         NaN         NaN   
1997-01-01  6451896.0         NaN         NaN         NaN         NaN   

                   72          84          96          108         120  
1988-01-01  10814576.0  10994014.0  11091363.0  11171590.0  11203949.0  
1989-01-01  12118790.0  12311629.0  12434826.0  12492899.0         NaN  
1990-01-01  13238667.0  13452993.0  13559557.0         NaN         NaN  
1991-01-01  13447481.0  13642414.0         NaN         NaN         NaN  
1992-01-01  14347271.0         NaN         NaN         NaN         NaN  
1993-01-01         NaN         NaN         NaN         NaN         NaN  
1994-01-01         NaN         NaN         NaN         NaN         NaN  
1995-01-01         NaN         NaN         NaN         NaN         NaN  
1996-01-01         NaN         NaN         NaN         NaN         NaN  
1997-01-01         NaN         NaN         NaN         NaN         NaN

Sometimes you just want the content copied to your clipboard, you can call `.to_clipboard()` and paste the result anywhere you like. 

In [31]:
clrd["CumPaidLoss"].sum().to_clipboard()

Other data I/O methods that you might want to know are `.to_json()` and `.to_pickle`. The inverse `chainladder.read_json()` and `chainladder.read_pickle()` are also available, but we won't explore them anymore here.

Now that we feel comfortable going in and out of `chainladder`, let's jump back in `chainladder` and explore some of the functions that an actuary often perform when working with triangles.

# Triangle Development

## Compute Loss Development Factors

Actuaries often spend lots of time trying to fine-tune their development factors, so let's explore the ways that `chainladder` can help us do that. `chainladder.Develompent()` is another helpful estimator, and has many useful attributes such as `.ldf_` or `.cdf_`.

Let's look at the dataset that we are already familiar with `clrd["CumPaidLoss"].sum()`.

In [32]:
cl.Development().fit(clrd["CumPaidLoss"].sum()).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.86453  1.230856  1.109122  1.055039  1.028329  1.015751  1.008899  1.005879  1.002897

In [33]:
cl.Development().fit(clrd["CumPaidLoss"].sum()).cdf_

12-Ult   24-Ult    36-Ult    48-Ult    60-Ult  72-Ult    84-Ult    96-Ult   108-Ult
(All)  2.854913  1.53117  1.243988  1.121597  1.063086  1.0338  1.017769  1.008792  1.002897

Remember `incr_to_cum()` from earlier? It works with development factors too!

In [34]:
cl.Development().fit(clrd["CumPaidLoss"].sum()).ldf_.incr_to_cum()

12-Ult   24-Ult    36-Ult    48-Ult    60-Ult  72-Ult    84-Ult    96-Ult   108-Ult
(All)  2.854913  1.53117  1.243988  1.121597  1.063086  1.0338  1.017769  1.008792  1.002897

You may have noticed that these attributes have a trailing underscore (\_). This is scikit-learn’s API convention, as its documentation states, “attributes that have been estimated from the data must always have a name ending with trailing underscore”. For example, the coefficients of some regression estimator would be stored in a `coef_` attribute after fit has been called. In summary, the trailing underscore in class attributes is a scikit-learn’s convention to denote that the attributes are estimated, or to denote that they are fitted attributes.

Now, you might ask, how are the averages calculated? By default, `chainladder.Development()` calculates these averages using all data and volume average the factors. Note that since the `.average` attribute is not estimated, it has no underscore following it.

In [35]:
cl.Development().fit(clrd["CumPaidLoss"].sum()).average

'volume'

Other averaging methods are `simple` and `regression`. The OLS `regression` estimate of development factors where the regression equation is Y = mX + 0.

In [36]:
simple_ldf = cl.Development(average="simple").fit(clrd["CumPaidLoss"].sum()).ldf_
simple_ldf

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.878245  1.233252  1.109947  1.055521  1.028566  1.015797  1.008927  1.005952  1.002897

In [37]:
regression_ldf = (
    cl.Development(average="regression").fit(clrd["CumPaidLoss"].sum()).ldf_
)
regression_ldf

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.851565  1.228516  1.108323  1.054585  1.028105  1.015706  1.008867  1.005806  1.002897

Remember, you can do simple arithmetic with any Triangle object.

In [38]:
simple_ldf - regression_ldf

12-24     24-36     36-48     48-60     60-72     72-84    84-96    96-108  108-120
(All)  0.02668  0.004736  0.001624  0.000936  0.000461  0.000092  0.00006  0.000146      NaN

We can also vary the `average` used for each age-to-age factors. Here we have 9 age-to-age factors, so we can supply an array of averages to use. Here we set the averages to `volume`, `simple`, `regression`, `volume`, `simple`, `regression`, `volume`, `simple`, `regression` for 12-24, 24-36, etc, respectively.

In [39]:
cl.Development(average=["volume", "simple", "regression"] * 3).fit(
    clrd["CumPaidLoss"].sum()
).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.86453  1.233252  1.108323  1.055039  1.028566  1.015706  1.008899  1.005952  1.002897

By default, `chainladder.Development()` sets `n_periods` as `-1` (use all data), which is the number of most recent periods to include in the calculation of averages. Let's try using only the most recent 3 periods.

In [40]:
cl.Development(average="simple", n_periods=3).fit(clrd["CumPaidLoss"].sum()).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.786207  1.214568  1.103199  1.052592  1.026814  1.015533  1.008927  1.005952  1.002897

## Discarding Problematic Link Ratios

From time to time, there might be certain data points that we may want to exclude from the calculation of loss development factors. For example, let's say we want to discard valuation 1991 (the diagonal).

In [41]:
cl.Development(drop_valuation="1991").fit(clrd["CumPaidLoss"].sum()).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.857588  1.228727  1.108023  1.053946  1.028329  1.015751  1.008899  1.005879  1.002897

Or that we want to drop a specific `origin` year's `age`.

In [42]:
cl.Development(drop=("1992", 24)).fit(clrd["CumPaidLoss"].sum()).ldf_

12-24    24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.86453  1.23059  1.109122  1.055039  1.028329  1.015751  1.008899  1.005879  1.002897

Or calculating the averages using the Olympic Average method, discarding the highest or lowest, or the highest or lowest n factors.

In [43]:
cl.Development(drop_high=[True, True, False, True], drop_low=[1, 2, 0, 3]).fit(
    clrd["CumPaidLoss"].sum()
).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.87613  1.237933  1.109122  1.057441  1.028329  1.015751  1.008899  1.005879  1.002897

## Setting Development Factors Manually

Sometimes, we want to manually set the age-to-age factors. `chainladder.DevelopmentConstant()` does exactly that. Note that we must also specify the `style` as `ldf` or `cdf`.

In [44]:
manual_patterns = {
    12: 1.8,
    24: 1.25,
    36: 1.10,
    48: 1.05,
    60: 1.03,
    72: 1.02,
    84: 1.01,
    96: 1.005,
    108: 1.002,
}
cl.DevelopmentConstant(patterns=manual_patterns, style="ldf").fit(
    clrd["CumPaidLoss"].sum()
).ldf_

12-24  24-36  36-48  48-60  60-72  72-84  84-96  96-108  108-120
(All)    1.8   1.25    1.1   1.05   1.03   1.02   1.01   1.005    1.002

Finally, before we go further, in `scikit-learn`, there are two types of estimators: transformers and predictors. A transformer transforms the input data (X) in some ways, and a predictor predicts a new value (or values, Y) by using the input data X.

`chainladder.Development()` and `chainladder.DevelopmentConstant()` are both transformers. The returned object is a means to create development patterns, which is used to estimate ultimates, but itself is not a IBNR estimation model, or a predictor.

In addition to `fit`, transformers come with the `transform` and `fit_transform` method. These will return a `chainladder.Triangle` object, but augment it with additional information for use in a subsequent IBNR model (a predictor). For example, `drop_high` can take an array of boolean variables, indicating if the highest factor should be dropped for each of the LDF calculation. 

Look at this example, calling `cl.Development().fit(clrd["CumPaidLoss"].sum()).ldf_` again actually doesn't give the transformed loss development factors that we had manually set, but using `fit_transform()` will actually modify the underlying attribute of the triangle, so we get the updated loss development factors.

In [45]:
cl.Development().fit(clrd["CumPaidLoss"].sum()).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120
(All)  1.86453  1.230856  1.109122  1.055039  1.028329  1.015751  1.008899  1.005879  1.002897

In [46]:
transformed_paid = cl.DevelopmentConstant(
    patterns=manual_patterns, style="ldf"
).fit_transform(clrd["CumPaidLoss"].sum())
transformed_paid.ldf_

12-24  24-36  36-48  48-60  60-72  72-84  84-96  96-108  108-120
(All)    1.8   1.25    1.1   1.05   1.03   1.02   1.01   1.005    1.002

One of the major benefits of `chainladder` is that it can handle several (or all) triangles simultaneously. While this can be a convenient shorthand, all these estimators will use the same assumptions across every triangle, as expected.

In [47]:
clrd_lob = cl.load_sample("clrd").groupby("LOB").sum()["CumPaidLoss"]
print("Fitting to " + str(len(clrd_lob.index)) + " industries simultaneously.")
cl.Development().fit_transform(clrd_lob).ldf_

Fitting to 6 industries simultaneously.


/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/core/pandas.py:415: FutureWarning: DataFrameGroupBy.dtypes is deprecated and will be removed in a future version. Check the dtypes on the base object instead
  if isinstance(self.groups.dtypes.index, pd.MultiIndex):
/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/core/pandas.py:427: FutureWarning: DataFrameGroupBy.dtypes is deprecated and will be removed in a future version. Check the dtypes on the base object instead
  index = pd.DataFrame(self.groups.dtypes.index)


Triangle Summary
Valuation:          2261-12
Grain:                 OYDY
Shape:         (6, 1, 1, 9)
Index:                [LOB]
Columns:      [CumPaidLoss]

## Correlation Tests

`chainladder` also has functionality to tests for possible violation of assumptions. The two main tests are:

1. The `valuation_correlation` test:
    * This test tests for the assumption of independence of `origin` years. In fact, it tests for correlation across calendar periods (diagonals), and by extension, origin periods (rows).
    * An additional parameter, `total`, can be passed, depending on if we want to calculate valuation correlation in total across all origins (`True`), or for each origin separately (`False`).
    * The test uses Z-statistic.
2. The `development_correlation` test:
    * This test tests for the assumption of independence of the chain ladder method that assumes that subsequent development factors are not correlated (columns).
    * The test uses T-statistic.

In [48]:
print(
    "Are valuation years correlated? I.e., are origins years correlated?",
    clrd["CumPaidLoss"]
    .sum()
    .valuation_correlation(p_critical=0.1, total=True)
    .z_critical.values,
)
print(
    "Are development periods coorelated?",
    clrd["CumPaidLoss"].sum().development_correlation(p_critical=0.5).t_critical.values,
)

Are valuation years correlated? I.e., are origins years correlated? [[False]]
Are development periods coorelated? [[ True]]


/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/core/correlation.py:231: RuntimeWarning: All-NaN slice encountered
  med = xp.nanmedian(


# Extending Development Patterns with Tail

## Setting Tail Factors Manually

Often, a tail factor is necessary to supplement our loss development factors, since our triangle is too "small". `chainladder.TailConstant()` is a useful estimator that has `tail` and `decay` that allow us to fine-tune our fail factors.
* `tail`: The constant to apply to all LDFs within a triangle object.
* `decay`: An exponential decay constant that allows for decay over future development periods. A decay rate of 0.5 sets the development portion of each successive LDF to 50% of the previous LDF.

In [49]:
cl.TailConstant(tail=1.005, decay=0.50).fit(clrd["CumPaidLoss"].sum()).cdf_

12-Ult    24-Ult    36-Ult    48-Ult    60-Ult    72-Ult    84-Ult    96-Ult   108-Ult  120-Ult   132-Ult
(All)  2.869188  1.538826  1.250208  1.127205  1.068402  1.038969  1.022858  1.013836  1.007911    1.005  1.002504

In [50]:
cl.TailConstant(tail=1.005, decay=0.50).fit(
    clrd["CumPaidLoss"].sum()
).cdf_ / cl.Development().fit(clrd["CumPaidLoss"].sum()).cdf_

12-Ult  24-Ult  36-Ult  48-Ult  60-Ult  72-Ult  84-Ult  96-Ult  108-Ult  120-Ult  132-Ult
(All)   1.005   1.005   1.005   1.005   1.005   1.005   1.005   1.005    1.005      NaN      NaN

## Compute Tail Factors

`chainladder.TailCurve()` is another class of tail transformers. Similar to the `chainladder.Development()` or `chainladder.TailConstant()` estimator, it comes with `fit`, `transform` and `fit_transform` methods. Also, like our `chainladder.Development()` estimator, you can define a tail in the absence of data or if you believe development will continue beyond your latest evaluation period.

Here, we can extend our development factors from 120 months to 144 months. 

In [51]:
clrd["CumPaidLoss"].sum().development.max()

120

In [52]:
tail = cl.TailCurve()
tail.fit(clrd["CumPaidLoss"].sum()).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120   120-132   132-144
(All)  1.86453  1.230856  1.109122  1.055039  1.028329  1.015751  1.008899  1.005879  1.002897  1.001245  1.001311

These extra twelve months (144 - 120, or one year) of development patterns are included, as it is typical for actuaries to track IBNR run-off over a 1-year time horizon from the valuation date. The tail extension is currently fixed at one year and there is no ability to extend it even further. However, a subsequent version of `chainladder` may address this issue.

Curve fitting takes selected development patterns and extrapolates them using either an `exponential` or `inverse_power` fit. In most cases, the `inverse_power` produces a thicker (more conservative) tail.

In [53]:
exp = cl.TailCurve(curve="exponential").fit(clrd["CumPaidLoss"].sum())
exp.tail_

120-Ult
(All)  1.002558

In [54]:
inv_power = cl.TailCurve(curve="inverse_power").fit(clrd["CumPaidLoss"].sum())
inv_power.tail_

120-Ult
(All)  1.026366

When fitting a tail, by default, all of the data will be used; however, we can specify which period of development patterns we want to begin including in the curve fitting process with `fit_period`, which takes a tuple of `start` and `stop` period. `None` can be used to ignore `start` or `stop`. For example, `(48, None)` will use development factors for age 48 and beyond. Alternatively, passing a list of booleans `[True, False, …]` will allow for excluding (False) any development patterns from fitting.

Patterns will also be generated for 100 periods beyond the end of the triangle by default, or we can specify how far beyond the triangle to project the tail factor to before dropping the age-to-age factor down to 1.0 using `extrap_periods`.

Note that even though we can extrapolate the curve many years beyond the end of the triangle for computational purposes, the resultant development factors will compress all ldf_ beyond one year into a single age-ultimate factor.

Let's ignore the first 3 development patterns for curve fitting but including the rest. Let's also allow our tail extrapolation to go 50 periods beyond the end of the triangle. Note that both `fit_period` and `extrap_periods` follow the `development_grain` of the underlying triangle being fit.

In [55]:
cl.TailCurve(fit_period=(36, None), extrap_periods=50).fit(
    clrd["CumPaidLoss"].sum()
).ldf_

12-24     24-36     36-48     48-60     60-72     72-84     84-96    96-108   108-120   120-132   132-144
(All)  1.86453  1.230856  1.109122  1.055039  1.028329  1.015751  1.008899  1.005879  1.002897  1.001603  1.001996

# IBNR Models

## Chainladder Model

Now that we have set and transformed the triangles' loss development factors, the IBNR estimators are the final stage in analyzing reserve estimates in the `chainladder` package. These estimators have a `predict` method as opposed to a `transform` method.

The most popular method, the chainladder method, can be called with `chainladder.Chainladder()`. The basic chainladder method is entirely specified by its development pattern selections. For this reason, the `chainladder.Chainladder()` estimator takes no additional assumptions, i.e. no additional arguments is needed.

In [58]:
cl_mod = cl.Chainladder().fit(clrd["CumPaidLoss"].sum())
cl_mod

Chainladder()

All IBNR models come with common attributes. First, the `.ultimate_` attribute, which gives the ultimate estimates from using the underlying model.

In [59]:
cl_mod.ultimate_

2261
1988  1.120395e+07
1989  1.252909e+07
1990  1.367877e+07
1991  1.388483e+07
1992  1.483220e+07
1993  1.595176e+07
1994  1.710361e+07
1995  1.742840e+07
1996  1.756851e+07
1997  1.841960e+07

Note that ultimates are measured at a valuation date way into the future. The library is extraordinarily conservative in picking this date, and sets it to December 31, 2261. This is set globally and can be viewed by referencing the `ULT_VAL` constant. This is a very common maximum time value across multiple python packages and holds no additional meaning other than that is commonly chosen.

In [62]:
cl.options.get_option("ULT_VAL")

'2261-12-31 23:59:59.999999999'

If for some reason, year `2261` is not far enough out the future for you, you can change this to whatever value you like.

In [67]:
cl.options.set_option("ULT_VAL", "2050-12-31 23:59:59.999999999")
print(cl.options.get_option("ULT_VAL"))
cl.options.set_option("ULT_VAL", "2261-12-31 23:59:59.999999999")  # Resetting it back

2050-12-31 23:59:59.999999999


Another commonly used attribute that is shared across all models is the `.ibnr_` attribute, which is calculated as the difference between `.ultimate_` and `.latest_diagonal`.

In [68]:
cl_mod.ibnr_

2261
1988           NaN
1989  3.618623e+04
1990  1.192173e+05
1991  2.424152e+05
1992  4.849332e+05
1993  9.466176e+05
1994  1.854279e+06
1995  3.418299e+06
1996  6.094599e+06
1997  1.196771e+07

Other attributes that actuaries might be interested in are the `.full_triangle_` and `.full_expectation_` attributes. While the `.full_expectation_` is entirely based on `.ultimate_` values and development patterns, the `.full_triangle_` is a blend of the existing triangle. These are useful for conducting an analysis of actual results vs model expectations.

In [69]:
cl_mod.full_triangle_

12            24            36            48            60            72            84            96            108           120           132           9999
1988  3577780.0  7.059966e+06  8.826151e+06  9.862687e+06  1.047470e+07  1.081458e+07  1.099401e+07  1.109136e+07  1.117159e+07  1.120395e+07  1.120395e+07  1.120395e+07
1989  4090680.0  7.964702e+06  9.937520e+06  1.109859e+07  1.176649e+07  1.211879e+07  1.231163e+07  1.243483e+07  1.249290e+07  1.252909e+07  1.252909e+07  1.252909e+07
1990  4578442.0  8.808486e+06  1.098535e+07  1.222900e+07  1.287854e+07  1.323867e+07  1.345299e+07  1.355956e+07  1.363927e+07  1.367877e+07  1.367877e+07  1.367877e+07
1991  4648756.0  8.961755e+06  1.115424e+07  1.240959e+07  1.309204e+07  1.344748e+07  1.364241e+07  1.376382e+07  1.384473e+07  1.388483e+07  1.388483e+07  1.388483e+07
1992  5139142.0  9.757699e+06  1.202798e+07  1.328948e+07  1.399282e+07  1.434727e+07  1.457325e+07  1.470293e+07  1.478937e+07  1.483220e+07  1.483220e+07  1.483220e+07
1993  5653379.0  1.059942e+07  1.295381e+07  1.429252e+07  1.500514e+07  1.543022e+07  1.567325e+07  1.581273e+07  1.590568e+07  1.595176e+07  1.595176e+07  1.595176e+07
1994  6246447.0  1.139496e+07  1.384576e+07  1.524933e+07  1.608863e+07  1.654441e+07  1.680499e+07  1.695454e+07  1.705421e+07  1.710361e+07  1.710361e+07  1.710361e+07
1995  6473843.0  1.161215e+07  1.401010e+07  1.553891e+07  1.639415e+07  1.685858e+07  1.712411e+07  1.727650e+07  1.737806e+07  1.742840e+07  1.742840e+07  1.742840e+07
1996  6591599.0  1.147391e+07  1.412273e+07  1.566383e+07  1.652595e+07  1.699412e+07  1.726178e+07  1.741539e+07  1.751777e+07  1.756851e+07  1.756851e+07  1.756851e+07
1997  6451896.0  1.202976e+07  1.480689e+07  1.642265e+07  1.732654e+07  1.781738e+07  1.809801e+07  1.825907e+07  1.836640e+07  1.841960e+07  1.841960e+07  1.841960e+07

In [70]:
cl_mod.full_expectation_

12            24            36            48            60            72            84            96            108           120           132           9999
1988  3.924445e+06  7.317247e+06  9.006475e+06  9.989278e+06  1.053908e+07  1.083764e+07  1.100834e+07  1.110630e+07  1.117159e+07  1.120395e+07  1.120395e+07  1.120395e+07
1989  4.388605e+06  8.182687e+06  1.007171e+07  1.117075e+07  1.178558e+07  1.211945e+07  1.231034e+07  1.241989e+07  1.249290e+07  1.252909e+07  1.252909e+07  1.252909e+07
1990  4.791310e+06  8.933543e+06  1.099590e+07  1.219580e+07  1.286704e+07  1.323155e+07  1.343996e+07  1.355956e+07  1.363927e+07  1.367877e+07  1.367877e+07  1.367877e+07
1991  4.863486e+06  9.068117e+06  1.116154e+07  1.237951e+07  1.306087e+07  1.343087e+07  1.364241e+07  1.376382e+07  1.384473e+07  1.388483e+07  1.388483e+07  1.388483e+07
1992  5.195326e+06  9.686843e+06  1.192311e+07  1.322418e+07  1.395202e+07  1.434727e+07  1.457325e+07  1.470293e+07  1.478937e+07  1.483220e+07  1.483220e+07  1.483220e+07
1993  5.587475e+06  1.041802e+07  1.282308e+07  1.422235e+07  1.500514e+07  1.543022e+07  1.567325e+07  1.581273e+07  1.590568e+07  1.595176e+07  1.595176e+07  1.595176e+07
1994  5.990937e+06  1.117028e+07  1.374901e+07  1.524933e+07  1.608863e+07  1.654441e+07  1.680499e+07  1.695454e+07  1.705421e+07  1.710361e+07  1.710361e+07  1.710361e+07
1995  6.104703e+06  1.138240e+07  1.401010e+07  1.553891e+07  1.639415e+07  1.685858e+07  1.712411e+07  1.727650e+07  1.737806e+07  1.742840e+07  1.742840e+07  1.742840e+07
1996  6.153781e+06  1.147391e+07  1.412273e+07  1.566383e+07  1.652595e+07  1.699412e+07  1.726178e+07  1.741539e+07  1.751777e+07  1.756851e+07  1.756851e+07  1.756851e+07
1997  6.451896e+06  1.202976e+07  1.480689e+07  1.642265e+07  1.732654e+07  1.781738e+07  1.809801e+07  1.825907e+07  1.836640e+07  1.841960e+07  1.841960e+07  1.841960e+07

And of course, you can back test to see how close the actuals are compared to what the model thinks in the upper left side of our triangle.

In [76]:
residuals = cl_mod.full_expectation_ - cl_mod.full_triangle_
residuals[residuals.valuation < clrd.valuation_date]

12             24             36             48            60            72            84            96   108
1988  346664.831949  257280.559690  180324.113059  126591.261484  64380.299520  23064.221102  14325.313154  14937.996844  NaN
1989  297924.747991  217984.817928  134187.244310   72162.484363  19089.590029    661.634062  -1289.003906 -14937.996844  NaN
1990  212868.212760  125057.385629   10556.379469  -33204.396438 -11503.761841  -7114.895903 -13036.309247           NaN  NaN
1991  214729.726101  106361.821952    7299.869403  -30079.911240 -31168.757261 -16610.959261           NaN           NaN  NaN
1992   56183.954712  -70855.995074 -104876.441230  -65307.098905 -40797.370446           NaN           NaN           NaN  NaN
1993  -65904.023231 -181406.235367 -130736.003765  -70162.339263           NaN           NaN           NaN           NaN  NaN
1994 -255509.812661 -224675.692141  -96755.161245            NaN           NaN           NaN           NaN           NaN  NaN
1995 -369139.883519 -229746.662616            NaN            NaN           NaN           NaN           NaN           NaN  NaN
1996 -437817.754101            NaN            NaN            NaN           NaN           NaN           NaN           NaN  NaN

With this, we can also force the IBNR run-off of future periods, let's say we want the next three years'.

In [79]:
cl_mod.full_triangle_.dev_to_val().cum_to_incr().loc[..., "1998":"2000"]

1998          1999          2000
1988           NaN           NaN           NaN
1989  3.618623e+04           NaN           NaN
1990  7.971060e+04  3.950674e+04           NaN
1991  1.214019e+05  8.091135e+04  4.010186e+04
1992  2.259778e+05  1.296853e+05  8.643201e+04
1993  4.250811e+05  2.430349e+05  1.394741e+05
1994  8.393079e+05  4.557755e+05  2.605840e+05
1995  1.528808e+06  8.552461e+05  4.644305e+05
1996  2.648819e+06  1.541098e+06  8.621217e+05
1997  5.577860e+06  2.777139e+06  1.615756e+06

Most of the above methods from the `chainladder.Chainladder()` model apply to actuarial models inside `chainladder`, which we will not repeatedly demonstrate. 

Let's look at another model, the `chainladder.ExpectedLoss()` model, which is when we know the ultimate loss already (but then why would we estimate ultimate losses?). The Expected Loss model requires one input assumption, the `aprior`, which is a scalar multiplier that will be applied to an exposure vector, that will produce an a priori ultimate estimate vector that we can use for the model.

Let's assume that our `aprior` is 80% of the scalar multiplier, and that this multiplier should be applied to `clrd["EarnedPremDIR"]`. But first, let's revisit `clrd["EarnedPremDIR"].sum()`.

In [87]:
clrd["EarnedPremDIR"].sum()

12          24          36          48          60          72          84          96          108         120
1988  14759891.0  14759891.0  14759891.0  14759891.0  14759891.0  14759891.0  14759891.0  14759891.0  14759891.0  14759891.0
1989  16251494.0  16251494.0  16251494.0  16251494.0  16251494.0  16251494.0  16251494.0  16251494.0  16251494.0         NaN
1990  17967080.0  17967080.0  17967080.0  17967080.0  17967080.0  17967080.0  17967080.0  17967080.0         NaN         NaN
1991  19662971.0  19662971.0  19662971.0  19662971.0  19662971.0  19662971.0  19662971.0         NaN         NaN         NaN
1992  21208358.0  21208358.0  21208358.0  21208358.0  21208358.0  21208358.0         NaN         NaN         NaN         NaN
1993  22951940.0  22951940.0  22951940.0  22951940.0  22951940.0         NaN         NaN         NaN         NaN         NaN
1994  24758613.0  24758613.0  24758613.0  24758613.0         NaN         NaN         NaN         NaN         NaN         NaN
1995  26121518.0  26121518.0  26121518.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN
1996  26810956.0  26810956.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN
1997  27076444.0         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN         NaN

## Expected Loss Model

Remember that `chainladder.ExpectedLoss()` applies the scaler to a vector, and not a triangle. But we also see that the premium does not develop over time, so we can just get any vector we want. With that said, we will use the `.latest_diagonal` premium vector.

In [84]:
cl.ExpectedLoss(apriori=0.80).fit(
    clrd["CumPaidLoss"].sum(), sample_weight=clrd["EarnedPremDIR"].latest_diagonal.sum()
).ultimate_

2261
1988  11807912.8
1989  13001195.2
1990  14373664.0
1991  15730376.8
1992  16966686.4
1993  18361552.0
1994  19806890.4
1995  20897214.4
1996  21448764.8
1997  21661155.2

A very common question that one might ask is what is the difference between:

```python
cl.ExpectedLoss(apriori=0.80).fit(..., sample_weight=weight_vector)
```

versus 

```python
cl.ExpectedLoss(apriori=1.00).fit(..., sample_weight=weight_vector*0.80)
```

Here is where `chainladder` follows `scikit-learn`'s implementation philosophy closely. The `apriori=...` inside the estimator, `chainladder.ExpectedLoss()` is a model parameter, whereas the inputs inside `fit()` are the data that the model will be applied to. With that said, only the first code block's syntax is theoretically correct, because in the second scenario, we are technically modifying our data, while not using the model assumption (i.e. `apriori`) correctly, even though they will yield identical results.

Let's apply the same model to `clrd["IncurLoss"].sum()` just to make sure that we get the exact same ultimates.

In [91]:
cl.ExpectedLoss(apriori=0.80).fit(
    clrd["IncurLoss"].sum(), sample_weight=clrd["EarnedPremDIR"].latest_diagonal.sum()
).ultimate_

/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/utils/weighted_regression.py:58: RuntimeWarning: Mean of empty slice
  xp.nansum(w * x * y, axis) - xp.nansum(x * w, axis) * xp.nanmean(y, axis)
/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/utils/weighted_regression.py:62: RuntimeWarning: Mean of empty slice
  intercept = xp.nanmean(y, axis) - slope * xp.nanmean(x, axis)


2261
1988  11807912.8
1989  13001195.2
1990  14373664.0
1991  15730376.8
1992  16966686.4
1993  18361552.0
1994  19806890.4
1995  20897214.4
1996  21448764.8
1997  21661155.2

## Bornhuetter-Ferguson

The `chainladder.BornhuetterFerguson()` estimator is another method having many of the same attributes as the `chainladder.Chainladder()` estimator. It comes with one input assumption, the a priori (`apriori`), a scalar multiplier that will be applied to an exposure vector, which will produce an a priori ultimate estimate vector that we can use for the model, which works exactly like `chainladder.ExpectedLoss()`.

In [92]:
cl.BornhuetterFerguson(apriori=0.80).fit(
    clrd["IncurLoss"].sum(), sample_weight=clrd["EarnedPremDIR"].latest_diagonal.sum()
).ultimate_

/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/utils/weighted_regression.py:58: RuntimeWarning: Mean of empty slice
  xp.nansum(w * x * y, axis) - xp.nansum(x * w, axis) * xp.nanmean(y, axis)
/Users/kennethhsu/opt/anaconda3/lib/python3.9/site-packages/chainladder/utils/weighted_regression.py:62: RuntimeWarning: Mean of empty slice
  intercept = xp.nanmean(y, axis) - slope * xp.nanmean(x, axis)


2261
1988  1.139698e+07
1989  1.272225e+07
1990  1.389113e+07
1991  1.407756e+07
1992  1.497869e+07
1993  1.593474e+07
1994  1.690885e+07
1995  1.686063e+07
1996  1.630346e+07
1997  1.582525e+07

## Benktander

The `chainladder.Benktander()` method is similar to the Bornhuetter-Ferguson method, but allows for the specification of one additional assumption, `n_iters`, the number of iterations to recalculate the ultimates. The Benktander method generalizes both the Bornhuetter-Ferguson and the Chainladder estimator through this assumption.

- When `n_iters = 1`, the result is equivalent to the Bornhuetter-Ferguson estimator.
- When `n_iters` is sufficiently large, the result converges to the Chainladder estimator.